### DISCUSSION TITLE
SQLite in static image

### DISCUSSION THREAD

**Discussion by freebuu** [php/frankenphp#839] [Q&A] [answered]
Hello there! I think, this is a stupid question, but i cannot figure out how to use SQLite database with static binary.

What i want to achieve: 
- Build static executable with laravel 11 project which uses SQLite - **done**
- Pack this executable in docker image - **done**
- Start service with this image in docker-compose (idea - move some dev helper functional from main code into serparate service) - **done**
- Persist data produced by this service in SQLite - **fail**

SQLite database file stored somewhere in executable and i cannot figure out how can i "mount" them to local filesystem/docker storage/etc. Or how can i open external sqlite file by this executable.

Thanks!

**Comment by withinboredom** [comment]
Are you asking how you can load a SQLite db, then store those changes in the executable?

I'm not sure such a thing is possible, or understand why you would want to do that. Can you clarify what you mean?

**Comment by nikspyratos** [comment]
Are you using a framework by any chance? Any DB adapter like SQLite should work in the same way as eg Laravel's `storage` directory expectations - it'd look by default for `/yourAppPath/database/database.sqlite`.

**Comment by freebuu** [comment]
Yes, I didn't write clearly enough.

We have a laravel project that uses SQLite. The database file (`database.sqlite`) is located in the project structure (./storage/sqlite.database, ./database/database.sqlite - it doesn’t matter)

Compile this project into a static binary file - `RUN EMBED=dist/app/ ./build-static.sh`

Run the resulting binary file - `./frankenphp-linux-x86_64 php-server`

Everything works - the question is where is this `database.sqlite` for the running application?

**Reply by withinboredom** [reply]
It gets uncompressed into a temporary folder to run the PHP code, wherever your OS is configured (e.g., `/tmp` on Linux). If you are wanting changes to the database to survive reboots, you should do some installation on first run and store it somewhere more permanent for the given OS/user (e.g., `~/.config/` on Linux).

**Reply by freebuu** [reply]
Thanks! This is exactly what I need - to tell the compiled application which database file to use. Do you have a link to documentation about this? Or a phrase that I need to google - I rarely encounter non-interpretable code execution and am a little confused :)

**Reply by withinboredom** [reply, marked as answer]
It will do whatever you program it to do. I'm not sure what you mean.

> I rarely encounter non-interpretable code execution and am a little confused

This is just PHP, there's nothing fancy going on. To help your mental model out a bit, here's what is going on:

1. When you build the application, your PHP code is "zipped" up and appended to the executable. The executable is told the exact memory address of your "zipped" up code.
2. When you run the application, it sees the memory address and "unzips" your PHP code to a temporary folder.
3. The application configures the server root to this folder and runs your code.

What you do and how you do it in step 3, is all 100% you (and whatever framework/libraries you use).

**Reply by freebuu** [reply]
Figured out about `/tmp`, thank you! 
`embedded PHP app 📦     {"path": "/tmp/frankenphp_e88843b5e114759501e9d32af0b3bb29"}`

Then - can I specify the path where the application will be unpacked? Or, ideally, the path where the application needs to read a specific file (/tmp/frankenphp_***/storage/sqlite.database => /app/sqlite.database)

**Reply by freebuu** [reply]
@withinboredom understand all, this is really stupid question. Thanks for your comment, everything became clear!

**Reply by withinboredom** [reply]
> Then - can I specify the path where the application will be unpacked?

That's getting into installer territory, which FrankenPHP is not (and I don't think it has plans to ever be). What you can do, however, is have your script be an installer, and if the installation isn't configured yet, install it. (see WordPress's famous ["5 minute installer"](https://winningwp.com/how-to-install-wordpress-the-famous-five-minute-install-made-simple/#:~:text=3.%20Running%20the,WordPress%20installation%20screen%3A) for some inspiration).

**Reply by freebuu** [reply]
Yes, i understand the trick about unzip and how compiled binary works with filesystem, everything was very simple.

**Comment by dunglas** [comment]
As you're using Laravel, you can use the `LARAVEL_STORAGE_PATH` environment variable to change the storage path: https://frankenphp.dev/docs/laravel/#changing-the-storage-path
